kube-runtime/src/lease.rs

```diff
@@ -0,0 +1,279 @@
+                }
+            }
+        }
+    }
+
+    #[tracing::instrument(skip(self, now))]
+    async fn try_acquire(&self, now: DateTime<Utc>) -> Result<(), TryAcquireError> {
+        let mut entry = self
+            .api
+            .entry(&self.name)
+            .await
+            .map_err(AcquireError::Get)
+            .map_err(TryAcquireError::Acquire)?
+            .or_insert(Lease::default);
+        let lease = entry.get_mut().spec.get_or_insert_with(LeaseSpec::default);
+        let lease_state = self.state(lease);
+
+        if let LeaseState::HeldByOther {
+            ref holder,
+            expires_at,
+        } = lease_state
+        {
+            if expires_at > now {
+                return Err(TryAcquireError::Conflict {
+                    holder: holder.clone(),
+                    expires_at,
+                });
+            }
+        }
+
+        if !matches!(lease_state, LeaseState::HeldBySelf { .. }) {
+            lease.holder_identity = Some(self.identity.clone());
+            lease.acquire_time = Some(MicroTime(now));
+            *lease.lease_transitions.get_or_insert(0) += 1;
+        }
+        lease.renew_time = Some(MicroTime(now));
+        lease.lease_duration_seconds = Some(self.lease_duration_secs);
+
+        entry
+            .commit()
+            .await
+            .map_err(AcquireError::Commit)
+            .map_err(TryAcquireError::Acquire)?;
+        Ok(())
+    }
+
+    #[tracing::instrument(skip(self))]
+    async fn release(&self) -> Result<(), ReleaseError> {
+        let mut entry = self
+            .api
+            .entry(&self.name)
+            .await
+            .map_err(ReleaseError::Get)?
+            .or_insert(Lease::default);
+        let lease = entry.get_mut().spec.get_or_insert_with(LeaseSpec::default);
+        match self.state(lease) {
+            LeaseState::Unheld => Ok(()),
+            LeaseState::HeldByOther { holder, .. } => Err(ReleaseError::AlreadyStolen { holder }),
+            LeaseState::HeldBySelf { .. } => {
+                lease.holder_identity = None;
+                lease.acquire_time = None;
+                lease.renew_time = None;
+                lease.lease_duration_seconds = None;
+                *lease.lease_transitions.get_or_insert(0) += 1;
+                entry.commit().await.map_err(ReleaseError::Commit)?;
+                Ok(())
+            }
+        }
+    }
+
+    fn state(&self, lease: &LeaseSpec) -> LeaseState {
+        let lease_duration = Duration::seconds(lease.lease_duration_seconds.unwrap_or(0).into());
+        let last_renewal = lease.renew_time.as_ref().map_or(MIN_DATETIME, |dt| dt.0);
+
+        match &lease.holder_identity {
+            None => LeaseState::Unheld,
+            Some(holder) if holder == &self.identity => LeaseState::HeldBySelf {
+                expires_at: last_renewal + lease_duration,
+                renew_at: last_renewal + lease_duration / 2,
```

Comment:
so this is a renewal at half of the lease duration. this makes sense as an initial default, but this is configurable in client-go. I think they by default they waited 80% of the duration before renewing IIRC.

EDIT: they also have a bunch of [jitter logic on their retries](https://github.com/kubernetes/client-go/blob/8e46da3fd12b54a2d1726fc9a9438ff8fc38494c/tools/leaderelection/leaderelection.go#L249-L260) :thinking:

Reply:
Definitely open for thoughts on this, this is fairly placeholdery. We also need to have some clock skew compensation between `LeaseState::HeldBySelf`'s `expires_at` and `HeldByOther`'s. Keeping this discussion open for now.